telemetry/Tracks/class-tracks-client.php

```diff
@@ -0,0 +1,187 @@
+
+/**
+ * Handles all operations related to the Tracks pixel.
+ */
+class Tracks_Client {
+	/**
+	 * Base URL of the Tracks pixel.
+	 */
+	protected const PIXEL_BASE_URL = 'https://pixel.wp.com/t.gif';
+
+	/**
+	 * Class singleton instance.
+	 *
+	 * @var ?Tracks_Client
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Events queued to be sent to the Tracks pixel.
+	 *
+	 * @var array<Tracks_Event>
+	 */
+	protected $events = array();
+
+	/**
+	 * Constructor.
+	 */
+	private function __construct() {
+		// Track events asynchronously (inject pixels in the footer).
+		add_action( 'admin_footer', array( $this, 'render_tracking_pixels' ) );
+		// Synchronously track any remaining events that could not be added to
+		// the footer.
+		add_action( 'shutdown', array( $this, 'record_remaining_events' ) );
+	}
+
+	/**
+	 * Instantiates the singleton.
+	 *
+	 * @return Tracks_Client
+	 */
+	public static function instance(): Tracks_Client {
+		if ( is_null( static::$instance ) ) {
+			static::$instance = new Tracks_Client();
+		}
+
+		return static::$instance;
+	}
+
+	/**
+	 * Enqueues an event to be recorded asynchronously.
+	 *
+	 * @param Tracks_Event $event The event to record.
+	 * @return bool|WP_Error True if the event was enqueued for recording.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function record_event_asynchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		static::instance()->events[] = $event;
+
+		return true;
+	}
+
+	/**
+	 * Records an event synchronously (using a GET request).
+	 *
+	 * @param Tracks_Event $event The event to record.
+	 * @return bool|WP_Error True if recording the event succeeded.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if any error occurred.
+	 */
+	public function record_event_synchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		$pixel_url = static::instance()->generate_pixel_url( $event );
+
+		if ( null === $pixel_url ) {
+			return new WP_Error(
+				'invalid_pixel',
+				'cannot generate tracks pixel for given input',
+				array( 'status' => 400 )
+			);
+		}
+
+		// Add the Request Timestamp and URL terminator just before the HTTP
+		// request.
+		$pixel_url .= '&_rt=' . static::build_timestamp() . '&_=_';
+
+		$request = wp_safe_remote_get(
+			$pixel_url,
+			array(
+				'blocking'    => false,
+				'redirection' => 2,
+				'httpversion' => '1.1',
+				'timeout'     => 1,
+			)
+		);
+
+		if ( is_wp_error( $request ) ) {
+			return $request;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Outputs a Tracks pixel for every registered event.
+	 */
+	public function render_tracking_pixels(): void {
```

Comment:
@hanifn How are the actual pixel-style tracking pixels used? I'm more used to the server-side only pixel requests, so just want to understand. Is it:

1. Detect I'm on a page where I want to track something e.g. `$pagenow === 'edit.php'`
2. Call `register_events( [ new Tracks_Event( 'edit_loaded', ... ) ] )` during page loads
3. Pixels automatically fire off on browser render

Is that correct?

Reply:
@alecgeatches Yes that is about how it works although the `register_events` method no longer exist now. Any events that are triggered after the pixel rendered will be handled during the `shutdown` action by the `record_remaining_events` method